### PR TITLE
Layers/Resources Split (Part 1)

### PR DESCRIPTION
**Pull request by jridgewell**
**Can you feel the excitement?!?!**

I bit off way more than I can chew with the first pass of this, so I scrapped it and I'm going to be sending this out piecemeal. This one moves only measurement and the cached layout rectangle to the newly created `Layers` service.

### CODE REVIEW

File: test/functional/test-resources.js [hunk 1]
```diff
@@ -17,10 +17,15 @@
 import {AmpDocSingle} from '../../src/service/ampdoc-impl';
 import {Resources} from '../../src/service/resources-impl';
 import {Resource, ResourceState} from '../../src/service/resource';
+import {LayerElement} from '../../src/service/layers-impl';
 import {VisibilityState} from '../../src/visibility-state';
 import {layoutRectLtwh} from '../../src/layout-rect';
 import * as sinon from 'sinon';
 
+function stupidMutateLayoutBox(resource, rect) {
```
Comment (choumx) [inline test/functional/test-resources.js]: 😆  `setLayerElementLayoutBox`

File: src/service/resource.js [hunk 2]
```diff
@@ -441,12 +402,7 @@ export class Resource {
    */
   completeCollapse() {
     toggle(this.element, false);
-    this.layoutBox_ = layoutRectLtwh(
-        this.layoutBox_.left,
-        this.layoutBox_.top,
-        0, 0);
-    this.isFixed_ = false;
-    this.element.updateLayoutBox(this.layoutBox_);
+    this.requestMeasure();
```
Comment (choumx) [inline src/service/resource.js]: Is a resource pass guaranteed to be scheduled at this point?

File: src/service/resource.js [hunk 3]
```diff
@@ -296,7 +286,8 @@ export class Resource {
 
     if (this.hasBeenMeasured()) {
       this.state_ = ResourceState.READY_FOR_LAYOUT;
-      this.element.updateLayoutBox(this.layoutBox_);
+      // TODO
```
Comment (choumx) [inline src/service/resource.js]: ?

File: src/service/layers-impl.js [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+  moveLayoutRect,
+} from '../layout-rect';
+import {computedStyle} from '../style';
+import {registerServiceBuilderForDoc} from '../service';
+import {viewportForDoc} from '../services';
+import {dev} from '../log';
+
+const LAYER_PROP_ = '__AMP__LAYER';
+
+export class Layers {
```
Comment (choumx) [inline src/service/layers-impl.js]: Description of class?

File: test/functional/test-resources.js (hunk 1 above)
Comment (jridgewell) [inline test/functional/test-resources.js]: Nope. This is reaching into a private variable in another, unrelated class. I want to make this method go away when we redo the Resource tests.

File: src/service/resource.js (hunk 3 above)
Comment (jridgewell) [inline src/service/resource.js]: Whoops.

File: src/service/resource.js (hunk 2 above)
Comment (jridgewell) [inline src/service/resource.js]: Meaning?

File: src/service/layers-impl.js (hunk 4 above)
Comment (jridgewell) [inline src/service/layers-impl.js]: Done.

File: src/service/resource.js (hunk 2 above)
Comment (choumx) [inline src/service/resource.js]: The layout box will be updated on the next pass, but `requestMeasure` doesn't schedule a resource pass. With this change, we should make sure this method is only called when a resource pass is scheduled.

Comment (jridgewell) [inline src/service/resource.js]: I don't see an issue. A pass will always be scheduled, it's just how long it has to wait before it happens.

**Comment by dvoytenko** [comment]
(Reviewing)

File: src/service/layers-impl.js [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+   * @param {!AmpElement} element
+   */
+  add(element) {
+    new LayerElement(element, this);
+  }
+
+  /**
+   * @param {!AmpElement} unusedElement
+   */
+  remove(unusedElement) {
```
Comment (dvoytenko) [inline src/service/layers-impl.js]: Why not implemented?

File: src/service/layers-impl.js [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    return this.viewport_;
+  }
+
+
+  /****************************
+   * Ferrying to LayerElement *
+   ****************************/
+
+  /**
+   * @param {!AmpElement} element
```
Comment (dvoytenko) [inline src/service/layers-impl.js]: Indicate that measure context required.

File: src/service/layers-impl.js [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+   */
+  remeasure(element) {
+    return LayerElement.for(element).remeasure();
+  }
+
+  /**
+   * @param {!AmpElement} element
+   * @return {!../layout-rect.LayoutRectDef}
+   */
+  getPageLayoutBox(element) {
```
Comment (dvoytenko) [inline src/service/layers-impl.js]: Can we only have one layout box? It'd be local layout box...

File: src/service/layers-impl.js [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+   */
+  getAbsoluteLayoutBox(element) {
+    return LayerElement.for(element).getAbsoluteLayoutBox();
+  }
+
+  /**
+   * @param {!AmpElement} element
+   * @return {boolean}
+   */
+  isFixed(element) {
```
Comment (dvoytenko) [inline src/service/layers-impl.js]: Is `isFixed` still needed? Should we just expose `getLayerElement` instead?

File: src/service/layers-impl.js [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+   * scroll position (they move with the scroll).
+   * @return {!../layout-rect.LayoutRectDef}
+   */
+  getAbsoluteLayoutBox() {
+    if (!this.isFixed()) {
+      return this.layoutBox_;
+    }
+
+    const viewport = this.layers_.getViewport();
+    return moveLayoutRect(this.layoutBox_, viewport.getScrollLeft(),
```
Comment (dvoytenko) [inline src/service/layers-impl.js]: Let's discuss. I'd like to get rid of BCR here and only use `offsetTop/Left/etc` for all calculations. And by extension get rid of most of concepts of Viewport around here.

I think I understand what you are trying to do here for transition. But I wonder if we may not be better getting by designing layers just the way we want and finding a way to either work Resources class by experiment or such?

**Comment by choumx** [comment]
Closing. Please reopen if necessary.

**Pull request closed without merging**